File: FlexTools/Modules/ForOthers/DanMcLoy-NepaliToEnglishGloss/insert_ge.py
# Inserts a \ge field after every \d_nep field that contains
# a word glossed in nep_en_gloss.txt
# WARNING:  ANY EXISTING \ge FIELDS WILL BE DELETED!!!!


# Settings
nepFld = 'd_nep'    # Input field that contains Nepali.  (Not necessarily unique.) 
enFld = 'ge'        # Output field that will contain English
glossFile = 'nep_en_gloss.txt'  # File containing mapping of Nepali to English


import codecs
import re

def devEssence(s):
    """Given a Devanagari input string (Unicode), returns a simplified form in which
    phonologically insiginicant differences are standardized."""
    s = re.sub(u'इ', u'ई', s)
    s = re.sub(u'उ', u'ऊ', s)
    s = re.sub(u'ि', u'ी', s)
    s = re.sub(u'ु', u'ू', s)
    s = re.sub(u'श', u'स', s)
    s = re.sub(u'ष', u'स', s)
    s = re.sub(ur'[- \u200C\u200D]', '', s)
    return s

# Read glosses file
f = codecs.open(glossFile, encoding='utf-8')
geDict = dict()
for line in f:
    m = re.search(r'(.+)\t(.+)', line)
    if (m):
        ge = m.group(2)
        de = devEssence(m.group(1))
        if de in geDict:
            geDict[de] += " / " + ge
        else:
            geDict[de] = ge
print "Loaded", len(geDict), "glosses...\n"
f.close()


# SFM files to process
infileName = 'Lhomi Dictionary Project-sfm.txt'
outfileName = 'out.txt'

# Process SFM files
infile = codecs.open(infileName, encoding='utf-8')
outfile = codecs.open(outfileName, encoding='utf-8', mode='w+')
ct = 0
ctTry = 0
for line in infile:
    m = re.match(ur'\\'+nepFld+ur"\s+(\S.*)", line.strip("\r\n"))   #***
    if (m):
        de = devEssence(m.group(1))
        ctTry += 1
        if de in geDict:
            outfile.write(line)
            outfile.write(u"\\" + enFld + ' ' +geDict[de] + "\n")   #***
            ct += 1
        else:
            outfile.write(line)
    else:
        m = re.match(ur'\\'+enFld+ur"\s+(\S.*)", line)
        if (m):
            # Don't output original ge field.
            print "Deleting original \\", enFld, "field: ", m.group(1)
        else:
            outfile.write(line)
print "Processed", ctTry, "\\" + nepFld + " fields.\nMatched and inserted", ct, "glosses."

infile.close()          #***
outfile.close()         #***

File: FlexTools/Modules/ForOthers/DanMcLoy-NepaliToEnglishGloss/insert_ge.pl
# Inserts a \ge field after every \d_nep field that contains
# a word glossed in nep_en_gloss.txt
# WARNING:  ANY EXISTING \ge FIELDS WILL BE DELETED!!!!

use utf8;
use strict;

my $nepFld = 'd_nep';
my $enFld = 'ge';
my $glossFile = 'nep_en_gloss.txt';

# Prepare files for I/O
my $infile = 'Lhomi Dictionary Project-sfm.txt';
my $outfile = 'out.txt';
# my $infile = $ARGV[0];
# my $outfile = $ARGV[1];
# die "usage:\n\tperl $0 INFILE.txt OUTFILE.txt\n" unless $ARGV[1];
open (IN, "<:encoding(UTF-8)", $infile) or die "Can't open $infile\n";
open (OUT, ">:encoding(UTF-8)", $outfile) or die "Can't write to $outfile\n";
open (GLOSS, "<:encoding(UTF-8)", $glossFile) or die "Can't open $glossFile\n";
my (%ge, $ge, $de, $ctGe);
while (<GLOSS>) {
	if (/(.*)\t(.*)/) {
		$ge = $2;
		$de = devEssence($1);
		if (defined $ge{$de}) {
			$ge{$de} = " / $ge";
		} else {
			$ge{$de} = $ge;
			$ctGe++;
		}
	}
}
print "Loaded $ctGe glosses...\n";

# Process and output
my $ct = 0;
my $ctTry = 0;
while (<IN>) {
	if (/^\\$nepFld\s+(\S.*)/) {
		$de = devEssence($1);
		$ctTry++;
		if (defined $ge{$de}) {
			print OUT $_;
			print OUT "\\$enFld $ge{$de}\n";
			$ct++;
		} else {
			print OUT $_;
		}
	} elsif (/^\\$enFld\s+(\S.*)/) {
		warn "Deleting original \\$enFld field: $1\n";
		# don't output original ge field
	} else {
		print OUT $_;
	}
}
close IN;
close OUT;
print "Processed $ctTry \\$nepFld fields.\nMatched and inserted $ct glosses.\n";

sub devEssence {
	$de = $_[0];
	# $de =~ tr/इउकिकुशष/ईऊकीकूसस/;
	$de =~ s/इ/ई/g;
	$de =~ s/उ/ऊ/g;
	$de =~ s/ि/ी/g;
	$de =~ s/ु/ू/g;
	$de =~ s/श/स/g;
	$de =~ s/ष/स/g;
	$de =~ s/[\x{200c}\x{200d} \-]//g;
	return $de;
}